renderer.h

```diff
@@ -1,14 +1,30 @@
 #pragma once
 
 #include "indexbuffer.h"
-#include "shader.h"
 #include "renderutils.h"
+#include "shader.h"
+#include "texture.h"
 #include "vertexarray.h"
 
+#include <memory>
+
 class Renderer
 {
 private:
+    static std::unique_ptr<Renderer>sInstance;
+
+protected:
+    std::unique_ptr<Shader>mShader { nullptr };
+
 public:
-    void Draw(const VertexArray& va, const IndexBuffer& ib, const Shader& shader) const;
-    void Clear() const;
+    Renderer() = default;
```

Comment:
To be a true singleton, you should hide this as "private" and have the `instance` method instantiate the one true singleton. You want to enforce the fact that you will not ever have a second one of these things.